netcreateorg/netcreate-itest
# Feature: Provenance

This introduces a revamped "Provenance" tab.

To implement this, we've introduced another major revision to the template format, and are adding a `version` designator to gracefully handle migrations in the future.

![screenshot_1574](https://github.com/netcreateorg/netcreate-itest/assets/1403057/0b7d1e90-b996-4f8a-ad24-7dd77108737a)

## Breaking Changes

NOTE: the template format has changed, but the app should still gracefully load older projects.  You may need to manually update the project's template to add new Provenance functionality.  See [Working with older projects](#working-with-older-projects) below.

* Add Template Versioning
  - Templates now have a version designation.
  - The version should match the github Release version. Hopefully this will be less confusing than having separate version numbers.  The one caveat is that it's harder to track interim development/alpha versions prior to a release.
  - When opening projects, the template version (or lack of a template version) will allow the system to decide how to migrate or ignore data fields. e.g. previously built-in fields like `comments` will continue to be displayed and will not break in table views.
* Nodes and Edges recently added semi-built-in `comments` and `provenance` fields to the default template as placeholders.  These have now been removed in favor of the "real" implementation. d713ca1a6f85fe832f0114ef810042054fb0b903

## Significant Features
* The Provenance tab has been reformatted to show two sections: "PROVENANCE" and "HISTORY".
* Any fields that are designated `isProvenance` will be added to the "PROVENANCE" tab.
* Added a `createdBy` and `updatedBy` built-in fields 6fc8dee6eea91c9fd96a323745fabf91f4c74d37

## Minor Features
* The old semi-built-in fields`Provenance` and `Comments` have been removed from Node and Edge tables.  Any new "isProvenance" field will be displayed in the tables.

See also #37 and [Provenance Ideas](https://docs.google.com/document/d/174JMW5ktUbS3Chygu82dwMdtUlkSk6ef4XcWwnFxXSE/edit#heading=h.knf6zq2bqjv8)


--- 

# To Test

The easiest way to test this is to create a new project. (`_default.template.toml` has been updated, so any new project should use the new template schema).
1. `./nc.js --dataset=newtest`
2. Create a token and log in. (`ncMakeTokens('new','a','newtest',2)`)
3. Create a node
4. Savet eh node
5. Click on "Provenance" -- you should see "Info Source" and "HISTORY" sections with your login data as the created by and updated by author.

### Working with older projects
Older projects (e.g. any project created before May 23, 2024), will still open.  
* Any fields that have not been designated as `isProvenance` (in the template definition) will be displayed in the main "ATTRIBUTES" tab.  e.g. if you recently created a project, it probaby has a `provenance` and `comments` field defined.  These will show up in "ATTRIBUTES".
* The new Provenance tab should work, e.g.:
  - "HISTORY" wil show created by and updated by.
  - Any created/updated author that is missing will be displayed as "(not recorded)"
  - Since there aren't any fields designated as `isProvenance`, there won't be any fields available.

If you want to add provenance fields to an older project you have two options:
1. Create a new field, e.g. "Info Source", or
2. Add `isProvenance` to an existing field to have it appear in the "PROVENANCE" tab.  e.g. you can convert the old `provenance` field into an active provenance field by adding the `isProvenance` flag.


# To Do
- [x] `updated` field is not updating
- [x] custom *edge* provenance fields are not accepting input
- [x] Update template migration to support semantic versioning strings (e.g. `10.1` is NOT less than `2.0`)? -- won't do now.  See #217 
- [X] Date fields do not support historical dates -- won't do now.  See #207
- [X] Support filter by provenance dates -- won't do now.  See #130
